src/llvm-cpufeatures.cpp

```diff
@@ -0,0 +1,151 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+//
+// Lower intrinsics that expose subtarget information to the language. This makes it
+// possible to write code that changes behavior based on, e.g., the availability of
+// specific CPU features.
+//
+// The following intrinsics are supported:
+// - julia.cpu.have_fma.$typ: returns 1 if the platform supports hardware-accelerated FMA.
+//
+// Some of these intrinsics are overloaded, i.e., they are suffixed with a type name.
+// To extend support, make sure codegen (in intrinsics.cpp) knows how to emit them.
+//
+// XXX: can / do we want to make this a codegen pass to enable querying TargetPassConfig
+//      instead of using the global target machine?
+
+#include "llvm-version.h"
+
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/PassManager.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Target/TargetMachine.h>
+#include <llvm/Support/Debug.h>
+
+#include "julia.h"
+
+#define DEBUG_TYPE "cpufeatures"
+
+using namespace llvm;
+
+extern TargetMachine *jl_TargetMachine;
+
+// whether this platform unconditionally (i.e. without needing multiversioning) supports FMA
+Optional<bool> always_have_fma(Function &intr) {
+    auto intr_name = intr.getName();
+    auto typ = intr_name.substr(strlen("julia.cpu.have_fma."));
+
+#if defined(_OS_WINDOWS_)
+    // FMA on Windows is weirdly broken (#43088)
+    return false;
+#elif defined(_CPU_AARCH64_)
+    return typ == "f32" || typ == "f64";
+#else
+    (void)typ;
+    return {};
+#endif
+}
+
+bool have_fma(Function &intr, Function &caller) {
+    auto unconditional = always_have_fma(intr);
+    if (unconditional.hasValue())
+        return unconditional.getValue();
+
+    auto intr_name = intr.getName();
+    auto typ = intr_name.substr(strlen("julia.cpu.have_fma."));
+
+    Attribute FSAttr = caller.getFnAttribute("target-features");
+    StringRef FS =
+        FSAttr.isValid() ? FSAttr.getValueAsString() : jl_TargetMachine->getTargetFeatureString();
+
+    SmallVector<StringRef, 6> Features;
+    FS.split(Features, ',');
```

Comment:
From talking to @oscardssmith earlier this isn't quite sufficient to be correct, since we next also need to remove from the list any feature that wasn't present in the particular sysimg copy that we are using. Otherwise we are setting have_fma here, but fma support may actually be disabled in the places we are using it.

Reply:
I'm not sure I understand. This function will generally only use the feature string from the `target-features` function attribute, which will represent the features for this particular sysimg copy. Unless that attribute isn't set, in which case we can just use the native target's features. Doesn't that suffice?

Reply:
No, that would return incorrect answers at runtime for the `have_fma` function, since the attributes set on the function may a superset of the attributes that were available to other compilation units.